datameta/validation.py

```diff
@@ -0,0 +1,204 @@
+    # Make sure referenced file names are unique
+    ref_fname_counts = Counter(mdatrec.value for mdatrecs in mdat_fnames_obj.values() for mdatrec in mdatrecs)
+    errors += [ (mdatrec.metadataset, mdatrec.metadatum.name, "Filename occurs multiple times in metadata")
+            for ref_fname, count in ref_fname_counts.items() if count > 1
+            for mdatrec in mdat_fnames_obj[ref_fname]]
+
+    # Make sure the files' filenames and the referenced filenames match
+    errors += [ (db_file, None, "File included without reference in metadata") for db_file in db_files.values() if db_file.name not in ref_fnames.keys() ]
+    errors += [ (mdatrec.metadataset, mdatrec.metadatum.name, "Referenced file not provided") for ref_fname, mdatrec in ref_fnames.items() if ref_fname not in f_names_obj ]
+
+    return f_names_obj, ref_fnames, errors
+
+def validate_submission_uniquekeys(db, db_files, db_msets):
+    errors = []
+
+    # Submission unique keys (includes those that are globally unique)
+    keys_submission_unique  = [ md.name for md in db.query(MetaDatum).filter(or_(MetaDatum.submission_unique==True, MetaDatum.site_unique==True)) ]
+    # Globally unique keys
+    keys_site_unique        = [ md.name for md in db.query(MetaDatum).filter(MetaDatum.site_unique==True) ]
+
+    # Validate the set of metadatasets with regard to submission unique key constraints
+    for key in keys_submission_unique:
+        value_msets = defaultdict(list)
+        # Associate all values for that key with the metadatasets it occurs in
+        for db_mset in db_msets.values():
+            for mdatrec in db_mset.metadatumrecords:
+                if mdatrec.metadatum.name==key:
+                    value_msets[mdatrec.value].append(db_mset)
+        # Reduce to those values that occur in more than one metadatast
+        value_msets = { k: v for k, v in value_msets.items() if len(v) > 1 }
+        # Produce errrors
+        errors += [ (db_mset, key, "Violation of intra-submission unique constraint") for msets in value_msets.values() for db_mset in msets ]
+
+    # Validate the set of metadatasets with regard to site-wise unique key constraints
+    for key in keys_site_unique:
+        value_msets = defaultdict(list)
+        # Associate all values for that key with the metadatasets it occurs in
+        for db_mset in db_msets.values():
+            for mdatrec in db_mset.metadatumrecords:
+                if mdatrec.metadatum.name==key:
+                    value_msets[mdatrec.value].append(db_mset)
+
+        # Query the database for the supplied values
+        q = db.query(MetaDatumRecord)\
+                .join(MetaDataSet)\
+                .join(MetaDatum)\
+                .filter(and_(
+                    MetaDataSet.submission_id != None,
+                    MetaDatum.name==key,
```

Comment:
```suggestion
                    MetaDatum.name == key,
```
the `!=` in the line above is padded..